crates/swc_ecma_transforms_compat/src/es2017/async_to_generator.rs

```diff
@@ -44,61 +44,58 @@ struct Actual {
     extra_stmts: Vec<Stmt>,
 }
 
-/// TODO: VisitMut
 #[fast_path(ShouldWork)]
-impl Fold for AsyncToGenerator {
-    noop_fold_type!();
+impl VisitMut for AsyncToGenerator {
+    noop_visit_mut_type!();
 
-    fn fold_module_items(&mut self, n: Vec<ModuleItem>) -> Vec<ModuleItem> {
-        self.fold_stmt_like(n)
+    fn visit_mut_module_items(&mut self, n: &mut Vec<ModuleItem>) {
+        self.visit_mut_stmt_like(n);
     }
 
-    fn fold_stmts(&mut self, n: Vec<Stmt>) -> Vec<Stmt> {
-        self.fold_stmt_like(n)
+    fn visit_mut_stmts(&mut self, n: &mut Vec<Stmt>) {
+        self.visit_mut_stmt_like(n);
     }
 }
 
 impl AsyncToGenerator {
-    fn fold_stmt_like<T>(&mut self, stmts: Vec<T>) -> Vec<T>
+    fn visit_mut_stmt_like<T>(&mut self, stmts: &mut Vec<T>)
     where
-        T: StmtLike + FoldWith<Actual>,
-        Vec<T>: FoldWith<Self>,
+        T: StmtLike + VisitMutWith<Actual>,
+        Vec<T>: VisitMutWith<Self>,
     {
-        let stmts = stmts.fold_children_with(self);
+        stmts.visit_mut_children_with(self);
 
-        let mut buf = Vec::with_capacity(stmts.len());
+        let stmts_len = stmts.len();
 
-        for stmt in stmts {
+        for idx in 0..stmts_len {
             let mut actual = Actual {
                 in_object_prop: false,
                 in_prototype_assignment: false,
                 extra_stmts: vec![],
             };
-            let stmt = stmt.fold_with(&mut actual);
 
-            buf.extend(actual.extra_stmts.into_iter().map(T::from_stmt));
-            buf.push(stmt);
-        }
+            let stmt = &mut (stmts[idx]);
+            stmt.visit_mut_with(&mut actual);
 
-        buf
+            stmts.extend(actual.extra_stmts.into_iter().map(T::from_stmt));
```

Comment:
I think this will append `extra_stmts` to the `stmts`, while previous behavior will prepend each injected statement prior to stmt.

Reply:
Yes it does. I wrote in PR for the changes

> One notable changes in this PR is this change actually changed output bit, between order of generated fn. Since visit_mut doesn't return whole stmts vec now it appends extra_stmts each time it's being created which ends up emission order changes. Peeking babel's behavior, actually this aligns with babel's output and afaik this shouldn't cause runtime differences.

and as said above, this actually makes output emission aligned to what babel does, thought it'd acceptable.

Reply:
Thank you for your explanation. Then it's fine.